### PR TITLE
feat(plugins): add teardown functionality to session replay plugin

### PR DESCRIPTION
**Pull request by kwalker3690**
### Summary
This PR adds a teardown method to the session replay plugin, which is called when either `amplitude.remove()` or `amplitude.init()` is called for a subsequent time. This method removes event listeners from the window, stops recording, and flushes the event queue, sending all remaining events.

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  no

### CODE REVIEW

File: packages/plugin-session-replay-browser/src/session-replay.ts [hunk 1]
```diff
@@ -76,39 +76,46 @@ class SessionReplay implements SessionReplayEnrichmentPlugin {
...
-    const GlobalScope = getGlobalScope();
-    if (GlobalScope && GlobalScope.window) {
-      GlobalScope.window.addEventListener('blur', () => {
-        this.stopRecordingEvents && this.stopRecordingEvents();
-        this.stopRecordingEvents = null;
-      });
-      GlobalScope.window.addEventListener('focus', () => {
-        void this.initialize();
-      });
+    const globalScope = getGlobalScope();
```
Comment (kwalker3690) [inline packages/plugin-session-replay-browser/src/session-replay.ts:79]: This change was based on looking at the code for page view tracking plugin, to make the reference on globalScope look the same

File: packages/plugin-session-replay-browser/src/session-replay.ts [hunk 2]
```diff
@@ -76,39 +76,46 @@ class SessionReplay implements SessionReplayEnrichmentPlugin {
...
   }
 
+  blurListener = () => {
+    this.stopRecordingAndSendEvents();
+  };
+  focusListener = () => {
+    void this.initialize();
+  };
+
+  stopRecordingAndSendEvents(sessionId?: number) {
```
Comment (kwalker3690) [inline packages/plugin-session-replay-browser/src/session-replay.ts:97]: Consolidating logic here, and actually discovered a bug doing so - any time we call `stopRecordingEvents`, we have to set it to null afterwards to signal that recording can begin again

File: packages/plugin-session-replay-browser/src/session-replay.ts [hunk 3]
```diff
@@ -195,9 +202,9 @@ class SessionReplay implements SessionReplayEnrichmentPlugin {
     }
     this.stopRecordingEvents = record({
       emit: (event) => {
-        const GlobalScope = getGlobalScope();
-        if (GlobalScope && GlobalScope.document && !GlobalScope.document.hasFocus()) {
-          this.stopRecordingEvents && this.stopRecordingEvents();
```
Comment (kwalker3690) [inline packages/plugin-session-replay-browser/src/session-replay.ts:200]: This is where we weren't setting stopRecordingEvents to null

File: packages/plugin-session-replay-browser/src/session-replay.ts [hunk 4]
```diff
@@ -76,39 +76,46 @@ class SessionReplay implements SessionReplayEnrichmentPlugin {
...
+    this.stopRecordingAndSendEvents();
+  };
+  focusListener = () => {
+    void this.initialize();
+  };
+
+  stopRecordingAndSendEvents(sessionId?: number) {
+    this.stopRecordingEvents && this.stopRecordingEvents();
+    this.stopRecordingEvents = null;
+    const sessionIdToSend = sessionId || this.config.sessionId;
```
Comment (jackson-amplitude) [inline packages/plugin-session-replay-browser/src/session-replay.ts:100]: when would this be different?

**Review by jackson-amplitude — APPROVED** [review]
LGTM! Love the organized teardown logic.

**Review by kevinpagtakhan — APPROVED** [review]
lgtm!

File: packages/plugin-session-replay-browser/src/session-replay.ts (hunk 4 above)
Comment (kwalker3690) [reply packages/plugin-session-replay-browser/src/session-replay.ts:100]: It's a safety check for the end session event, I want to make sure we send events with the session id associated with the end session event, not the next session id that gets created when a start session event fires

**Pull request merged**